Bug fix - do not display empty bars when line.width is zero

archmoj opened the pull request:
Fixes #4047.
Empty `bar` and `funnel` rectangles should not be displayed when marker.line.width is set to zero.

[Codepen after](https://codepen.io/MojtabaSamimi/pen/bXbmba)

Also addressing #4057 for bars in this PR.
[Codepen after](https://codepen.io/MojtabaSamimi/pen/mNbYNL)

@plotly/plotly_js

etpinard commented on src/traces/bar/plot.js:
```diff
@@ -102,12 +102,23 @@ function plot(gd, plotinfo, cdModule, traceLayer, opts) {
+            // display zeros if line.width > 0
+            if(isBlank && shouldDisplayZeros) {
+                if(isHorizontal ? x1 - x0 === 0 : y1 - y0 === 0) {
+                    if(di.mlw || trace.marker.line.width > 0) {
```
It's unclear whether this means `di.mlw || (trace.marker.line.width > 0)` or `(di.mlw || trace.marker.line.width) > 0`. Both are equivalent in this scenario, but to me 

```js
if(di.mlw > 0 || trace.marker.line.width > 0)
```

or simply

```js
if(di.mlw || trace.marker.line.width)
```

would be better.

---

Moreover, I would prefer if we combined all three `if()` into one. I don't see the need to have :three: if statements here. Oh well, this part in non-blocking. I'll let @archmoj decide what's best.

archmoj replied:
@etpinard quick question:
I am wondering if `di.mlw` could be a negative number?

etpinard replied:
Probably, as we don't coerce `arrayOk` attributes. That logic must be in a `Drawing.style*` method (please double-check though), so at this stage in `plot.js` we should **not** assume that `di.mlw >= 0`. Good eye!

etpinard commented on test/image/mocks/bar_display_height_zero_only_line_width.json:
```diff
@@ -0,0 +1,340 @@
+            "type": "bar",
+            "marker": {
+                "line": {
+                    "width": [10, 0, 10, 0, 10]
```
Looks like this mock only has array `marker.line.width` value. Could you add one trace that test scaler `marker.line.width`? 

Maybe you can leave this mock alone and simply add another mock using https://codepen.io/MojtabaSamimi/pen/bXbmba , so that we also test the relative-stacked case.

etpinard commented:
Thanks for taking this one @archmoj - I made two minor comments.

archmoj commented:
@etpinard Thanks for the review!
This is now ready for the second round.
As I updated the PR description please note that the #4057 is also addressed for bars.

archmoj commented on src/traces/bar/plot.js:
```diff
@@ -130,8 +137,7 @@ function plot(gd, plotinfo, cdModule, traceLayer, opts) {
                 }
             } else {
-                lw = (di.mlw + 1 || trace.marker.line.width + 1 ||
-                    (di.trace ? di.trace.marker.line.width : 0) + 1) - 1;
```
@etpinard is there a need for this second line condition? i.e. <pre>di.trace ? di.trace.marker.line.width : 0</pre>.

etpinard replied:
Wow. Nice catch! That `di.trace` looks pretty odd. Maybe the old [toolpanel](https://github.com/plotly/streambed/tree/master/shelly/toolpanel/static/toolpanel/src/js) uses it. Would you wind trying to `grep` in the streambed directory?

archmoj replied:
![grep](https://user-images.githubusercontent.com/33888540/61416809-02c40280-a8c3-11e9-913a-88443d70f449.png)

etpinard replied:
and what if you grep for `Bar.plot`?

archmoj replied:
![grep_barPlot](https://user-images.githubusercontent.com/33888540/61417056-d957a680-a8c3-11e9-8a1b-a2cb7b7d2a49.png)

etpinard replied:
Great! I think we're good to :hocho: that `di.trace` condition.

etpinard commented on src/traces/bar/helpers.js:
```diff
@@ -64,3 +64,12 @@ exports.getValue = function(arrayOrScalar, index) {
+exports.getLineWidth = function(trace, di) {
+    var w =
+        (0 < di.mlw) ? di.mlw :
+        (0 < trace.marker.line.width) ? trace.marker.line.width : // Note: we don't want to get true & return an array when trace.marker.line.width is an array containing zeros.
```
We can mostly drop this `trace.marker.line.width` will never be negative after the defaults.

archmoj replied:
But it could be an array. For example if `line.width` is set to [0,1,2] in the first item di.mlw is zero.

etpinard replied:
True, so `Array.isArray(trace.marker.line.width` would be a better check on this line.

archmoj replied:
We also need to check for `null`s and other things. Let's keep the logic as is.

etpinard replied:
`trace.marker.line.width` should never be `null` at this stage.

Do you have an example of that?

archmoj replied:
You are right. Because this is coerced it shouldn't happen.

etpinard commented on src/traces/bar/helpers.js:
```diff
@@ -64,3 +64,12 @@ exports.getValue = function(arrayOrScalar, index) {
     return value;
 };
+
+exports.getLineWidth = function(trace, di) {
```
Bars don't have `line.width`, but still calling this `getMarkerLineWidth` would be better.

etpinard commented on src/lib/index.js:
```diff
@@ -459,10 +459,10 @@ lib.noneOrAll = function(containerIn, containerOut, attrList) {
     if(lib.isArrayOrTypedArray(traceAttr)) {
         var imax = Math.min(traceAttr.length, cd.length);
-        for(var i = 0; i < imax; i++) cd[i][cdAttr] = traceAttr[i];
+        for(var i = 0; i < imax; i++) cd[i][cdAttr] = isNumber ? +traceAttr[i] : traceAttr[i];
```
What we should really do here is: drop this new `isNumber` argument and call `isNumeric`.

```js
for(var i = 0; i < imax; i++) {
  var v = traceAttr[i];
  cd[i][cdAttr] = isNumerix(v) ? +v : v;
}
```

This might result in a performance drop, but this sounds like an acceptable change until we drop `gd.calcdata` all together.

archmoj replied:
What about keeping `isNumber` argument and only apply `isNumerix` on numeric fields e.g. `width` not `color`? That would help keep a better performance.

etpinard replied:
Yeah we could do that. But remember `marker.color` can have numeric items. I presume that they are handling elsewhere (probably somewhere in `component/colorscale/`).

---

Moreover, passing a plain boolean to a function like 

```js
mergeArray(markerLine.width, cd, 'mlw', true);
```

is considered a bad pattern (to many) as you can't guess what the function does by looking at the line-of-code of the caller (here, you would need to open `src/lib/index.js` to see what the `true` does)

Here are two "better" patterns:

```js
mergeArray(markerLine.width, cd, 'mlw', {coerceNumber: true});

// or
mergeArray(markerLine.width, cd, 'mlw', function(v) { return isNumeric(v) ? +v : v; });
```

Personally I prefer the latter (as it is more general) and it consistent with how [`Lib.fillArray`](https://github.com/plotly/plotly.js/blob/master/src/lib/index.js#L479-L487) and [`Lib.castOption`](https://github.com/plotly/plotly.js/blob/218a9195832aab94b1d8bd0ca4653c159d2f25ce/src/lib/index.js#L498-L512) work.

archmoj replied:
@etpinard thanks for the insight.
In https://github.com/plotly/plotly.js/pull/4056/commits/3ee7dbffe00cf5a899ee9858abe221e06e9e2e3f the second pattern you mentioned is applied to expand `mergeArray` functionalities.
Also in order to avoid requiring `fast-is-numeric` in `arrays_to_calcdata` steps in different traces, new functions were added to the `Lib`.

etpinard replied:
Adding `Lib.mergeArrayCastPositive` is a very good idea. But less not add `Lib.mergeArrayCastNumber` and `Lib.mergeArrayCastIfNumber` if you're not going to use them anywhere.

etpinard replied:
.... unless you're planning on fixing more `Lib.mergeArray` issues in this PR :smile:

archmoj replied:
Let's do that in a separate PR.
For now I cut them from the Lib in https://github.com/plotly/plotly.js/pull/4056/commits/579a1b55c248350de6c2d461c87ac0a05bd0eef9.

etpinard commented on src/lib/index.js:
```diff
@@ -459,13 +459,39 @@ lib.noneOrAll = function(containerIn, containerOut, attrList) {
-        for(var i = 0; i < imax; i++) cd[i][cdAttr] = isNumber ? +traceAttr[i] : traceAttr[i];
+        for(var i = 0; i < imax; i++) {
+            var v = traceAttr[i];
+            cd[i][cdAttr] = hasFn ? fn(v) : v;
```
I wonder what's faster 

```js
var hasFn = typeof fn === 'function';
for(var i = 0; i < N; i++) {
  arrayOut[i] = hasFn ? fn(arrayIn[i]) : arrayIn[i];
}
```

or 

```js
fn = typeof fn === 'function' ? fn : Lib.noop;
for(var i = 0; i < N; i++) {
  arrayOut[i] = fn(arrayIn[i]);
}
```

No need to change anything, I'm just thinking out-loud.

archmoj replied:
Interesting question. And I like the minimalist version with `Lib.noop`!
That may depend on the browser.
Having two loops may also be fast.
```js 
var i = 0;
if (typeof fn === 'function') {
  for(; i < N; ++i) arrayOut[i] = fn(arrayIn[i]);
} else {
  for(; i < N; ++i) arrayOut[i] = arrayIn[i];
}
```

etpinard commented:
Nicely done :dancer: 

Thanks for the fixes!
